src/env-inl.h

```diff
@@ -436,6 +521,52 @@ inline void Environment::SetTemplateMethod(v8::Local<v8::FunctionTemplate> that,
   function->SetName(name_string);  // NODE_SET_METHOD() compatibility.
 }
 
+inline void Environment::AddSubWorkerContext(WorkerContext* context) {
+  sub_worker_context_count_++;
+  sub_worker_context_list_dirty_ = true;
+  sub_worker_contexts()->PushBack(context);
+}
+
+inline void Environment::RemoveSubWorkerContext(WorkerContext* context) {
+  sub_worker_context_count_--;
+  sub_worker_context_list_dirty_ = true;
+  sub_worker_contexts()->Remove(context);
+}
+
+inline Environment::WorkerContextList* Environment::sub_worker_contexts() {
+  return &sub_worker_contexts_;
+}
+
+inline uv_mutex_t* Environment::ApiMutex() {
+  return is_worker_instance() ? worker_context()->api_mutex() : nullptr;
+}
+
+inline bool Environment::CanCallIntoJs() const {
+  return is_main_instance() ||
+         (is_worker_instance() && worker_context()->JsExecutionAllowed());
+}
+
+inline void Environment::Exit(int exit_code) {
+  if (is_main_instance())
+    exit(exit_code);
+  else
+    worker_context()->Exit(exit_code);
+}
+
+inline void Environment::ProcessNotifications() {
+  if (is_worker_instance())
+    WorkerContext::WorkerNotificationCallback(worker_context());
+
+  process_owner_notifications:
+  sub_worker_context_list_dirty_ = false;
+  for (WorkerContext* context : *sub_worker_contexts()) {
+    WorkerContext::OwnerNotificationCallback(context);
+    // The loop needs to be restarted if the list changed while in-loop.
+    if (sub_worker_context_list_dirty_)
+      goto process_owner_notifications;
```

Comment:
Maybe this is better handled by popping the contexts from the front, pushing them onto a temporary queue and splicing them back in afterwards.

The problem with the current approach is that it can result in approx. O(n * m) callbacks (with n = # of contexts, m = # of mutations.)